### PR TITLE
Hotfix265 dry mapper netcdf cf

### PR DESCRIPTION
**Pull request by mortenwh**
This should solve #265

### CODE REVIEW

**Comment by aleksandervines** [comment]
sub_filenames looks like a private method, hence I'd think the name should start with underscore.
As you changed the interface to this method, I would suggest you also change the name while you are at it :)

The smaller "public interface" nansat has, the easier it is to maintain backwards compatibility while still being able to refactor the code as much as possible.

**Comment by akorosov** [comment]
1. I agree with @aleksandervines 
2. Why did you remove gdal_dataset from input parameters? Now you have to repeat gdal.Open in several places. Time consuming and not very DRY.
3. If a netcdf-cf file doesn't have subdatasets (e.g. in case of one band only) sub_filenames will raise WrongMapperError. Why?
4. In sub_filenames_with_projection you loop over all subdatasets. That can be veery slow. And then you use only the first one. (and then open it again with gdal)

Suggested logic (take care of indentations!):
```
# Move sub_filenames() to VRT
class VRT():
    # ...
    def _get_sub_filenames(self, gdal_dataset):
        # Get filenames of subdatasets
        return [f[0] for f in gdal_dataset.GetSubDatasets()]


class Mapper(VRT):
def __init__(....):
    # ........ other stuff
    sub_filenames = self._get_sub_filenames(gdal_dataset)
    try:
        _init_from_subdataset_with_proj(gdal_dataset, gdal_metadata, sub_filenames)
    except NoSubDataSetWithProjError:
        _init_from_subdataset(gdal_dataset, gdal_metadata, sub_filenames)

def _init_from_subdataset_with_proj(gdal_dataset, gdal_metadata, sub_filenames):
    for sub_filename in sub_filenames:
        ds = gdal.Open(sub_filename)
        if ds.GetProjection():
            self._init_from_gdal_dataset(ds, metadata=gdal_metadata)
    raise NoSubDataSetWithProjError

def _init_from_subdataset(gdal_dataset, gdal_metadata, sub_filenames):
    # if no subdatasets (e.g., only one band and no time variable) use the input file alone
    if not  sub_filenames:
        ds = gdal_dataset
    else:
        ds = gdal.Open(sub_filenames[0])
    self._init_from_dataset_params(
                    x_size = ds.RasterXSize,
                    y_size = ds.RasterYSize,
                    geo_transform = ds.GetGeoTransform(),
                    projection = NSR().wkt,
                    metadata = gdal_metadata)

```

**Comment by akorosov** [comment]
Looks great now! One thing though:
In `def _create_empty_with_nansat_spatial_reference_WKT(self, gdal_dataset, gdal_metadata):`
try/except will always pass. Even if GDAL dataset has subdatasets you still can access RasterXSize, RasterYSize, GetGeoTransform.

I think it is better to 
```
fn = self._get_sub_filenames(gdal_dataset)
if  len(fn) == 0:
    subdataset = gdal_dataset
else:
    subdataset = gdal.Open(fn[0])

self._init_from_dataset_params(
                    x_size = subdataset.RasterXSize,
                    y_size = subdataset.RasterYSize,
                    geo_transform = subdataset.GetGeoTransform(),
                    projection = NSR().wkt,
                    metadata = gdal_metadata)
```

**Comment by mortenwh** [comment]
@akorosov - what is the status here? could you have a look?

**Comment by akorosov** [comment]
@mortenwh , have you fixed according to my comments? I think it is pending on you.

**Comment by mortenwh** [comment]
Yes, I have. See 4f7b225dae76f189700fb9f21769da9644083dcc

**Comment by akorosov** [comment]
@mortenwh , that was concerning the previous comment. And the last one (avoiding try/except and warning) was not fixed. I can do that myself now.

File: nansat/mappers/mapper_netcdf_cf.py [hunk 1]
```diff
@@ -194,7 +194,7 @@ class BreakI(Exception):
                 # Keep only desired bands (given in "bands" list)
                 try:
                     if bands:
-                        if not band_metadata.has_key('standard_name'):
+                        if 'standard_name' not in band_metadata.keys():
```
Comment (mortenwh) [inline nansat/mappers/mapper_netcdf_cf.py]: what's the reason for this change? is it faster?

File: nansat/mappers/mapper_netcdf_cf.py [hunk 2]
```diff
@@ -311,7 +311,7 @@ def _band_dict(self, subfilename, band_num, subds, band=None, band_metadata=None
                 bandName = band_metadata.get('dods_variable', '')
 
             # remove digits added by gdal when exporting to netcdf
-            if (len(bandName) > 0 and band_metadata.has_key('origin')
+            if (len(bandName) > 0 and 'origin' in band_metadata.keys()
```
Comment (mortenwh) [inline nansat/mappers/mapper_netcdf_cf.py]: same here..

File: nansat/mappers/mapper_netcdf_cf.py [hunk 3]
```diff
@@ -362,26 +362,18 @@ def _create_empty_with_nansat_spatial_reference_WKT(self, gdal_dataset, gdal_met
...
-        except Exception as e:
-            # issue warning to find out if this try-except clause is really needed...
-            warnings.warn('Main GDAL dataset does not contain geotransform or information ' \
-                    'about raster size: %s' %e.message)
-            fn = self._get_sub_filenames(gdal_dataset)
-            sub = gdal.Open(fn[0])
-            xs = sub.RasterXSize
-            ys = sub.RasterYSize
-            gt = sub.GetGeoTransform()
+        fn = self._get_sub_filenames(gdal_dataset)
```
Comment (mortenwh) [inline nansat/mappers/mapper_netcdf_cf.py]: this is fewer lines and looks ok, but still the call to `_get_sub_filenames` could be avoided by first trying to read the attributes of `gdal_dataset` itself...

File: nansat/mappers/mapper_netcdf_cf.py (hunk 1 above)
Comment (korvinos) [reply nansat/mappers/mapper_netcdf_cf.py]: method has_key was deprecated in Python 3
